crates/astria-sequencer/src/mempool/transactions_container.rs

```diff
@@ -159,6 +174,86 @@ impl PendingTransactionsForAccount {
     fn highest_nonce(&self) -> Option<u32> {
         self.txs.last_key_value().map(|(nonce, _)| *nonce)
     }
+
+    /// Removes and returns transactions that exceed the balances in `available_balances`.
+    fn find_demotables(
+        &mut self,
+        available_balances: &HashMap<IbcPrefixed, u128>,
+    ) -> Vec<TimemarkedTransaction> {
+        let mut split_at = 0;
+        let mut break_flag = false;
+        let mut available_balances = available_balances.clone();
+
+        for (nonce, tx) in &self.txs {
+            // ensure we have enough balance to cover inclusion
+            for (denom, cost) in tx.cost() {
+                match available_balances.entry(*denom) {
+                    hash_map::Entry::Occupied(mut entry) => {
+                        if cost <= entry.get() {
+                            // subtract cost
+                            let current_cost = entry.get_mut();
+                            *current_cost = current_cost.saturating_sub(*cost);
+                        } else {
+                            // not enough balance, do not include
+                            break_flag = true;
+                            break;
+                        }
+                    }
+                    hash_map::Entry::Vacant(_) => {
+                        // not enough balance, do not include
+                        if *cost != 0 {
+                            break_flag = true;
+                            break;
+                        }
+                    }
+                }
+            }
+
+            // break if we're out of funds for one since this is a `SequentialNonce` container
+            if break_flag {
+                break;
+            }
+            split_at = nonce.saturating_add(1);
+        }
+
+        // return all keys higher than split target
+        self.txs.split_off(&split_at).into_values().collect()
+    }
+
+    /// Returns remaining balances after accounting for costs of contained transactions.
+    ///
+    /// Note: assumes that the balances in `current_account_balance` are large enough
+    /// to cover costs for contained transactions. Will log an error if this is not true
+    /// but will not fail.
```

Comment:
why not return an error? what cases would that error happen, developer error or other?

Reply:
It'd be a code logic error. It is possible to have too high costs in pending after transaction re-costing or a different node including a different transaction for the same account's nonce in a block, but we only call `get_remaining_balances()` after we run the demotion logic on pending which should make the transaction costs covered.

Fraser and I thought it would be better to let the mempool continue running if somewhere else in the code there was a bug that would make the above assumptions untrue instead of having the mempool crash. The idea was that someone would see the error logs and we'd know we need to go fix something 😬